src/device.cpp

```diff
@@ -146,10 +146,20 @@ namespace D3D11On12
     //----------------------------------------------------------------------------------------------------------------------------------
     HRESULT Device::EnsureVideoDevice()
     {
-        if (!m_pVideoDevice)
+        try
+        {
+            if (!m_pVideoDevice)
+            {
+                m_pVideoDevice.reset(new VideoDevice(*this));
+                m_pVideoDevice->Initialize();
+            }
+        }
+        catch (_com_error& hrEx)

```

Comment:
I think this should be:
```c++
catch (...)
{
    m_pVideoDevice = nullptr;
    throw;
}
```

Since there could be `bad_alloc` or potentially other exceptions that would still leave things in a bad state. The caller already needs to expect exceptions, no need to convert to HRESULT here (in fact I'd consider going further and making the return code void since it's always `S_OK` or an exception).

Reply:
The caller (`HRESULT APIENTRY Device::RetrieveSubObject()` in device.cpp) currently just checks for SUCCEEDED and must return an HRESULT across the API boundary anyway, so if we don't convert to HRESULT here we'll just have to do it later anyway. Kinda feels like "6 of one, half a dozen of the other". For what it's worth, the rest of device.cpp doesn't use exceptions, so this is consistent at least.

Reply:
The RetrieveSubObject does translate all exceptions to HRESULT.  This is sneakily hidden inside the D3D11on12_DDI_ENTRYPOINT_START and D3D11on12_DDI_ENTRYPOINT_END_AND_RETURN_HR macros.

Reply:
Ah, that does ring a bell now that you mention it. Ignore my ramblings then 🙃